File: frontend/src/tipos/chatbot.ts
import type { Nullable } from './api';

// --- ENUMS (Coinciden con Backend Go) ---

export type TipoChatbot = 'ASISTENTE_IA' | 'WHATSAPP_BOT' | 'TELEGRAM_BOT' | 'CUSTOM';

// El backend valida "oneof=LIVE TEST", así que debemos enviar estos valores exactos.
export type EntornoAPIKey = 'LIVE' | 'TEST';

// --- INTERFACES PRINCIPALES ---

export interface Chatbot {
  id: string;
  tenant_id: string;
  nombre: string;
  tipo: TipoChatbot;
  descripcion: Nullable<string>;

  // Configuración IA (Nuevos campos v4)
  modelo_ia: Nullable<string>;       // ej: "gpt-4o", "claude-sonnet"
  prompt_sistema: Nullable<string>;  // Instrucciones base
  temperatura: Nullable<number>;     // 0.0 a 1.0
  max_tokens_respuesta: Nullable<number>;

  // Scopes / Permisos (Qué puede hacer este bot)
  puede_leer_reclamos: boolean;
  puede_responder: boolean;
  puede_cambiar_estado: boolean;
  puede_enviar_mensajes: boolean;
  puede_leer_metricas: boolean;

  // Reglas de Negocio
  requiere_aprobacion: boolean;      // Si true, respuestas quedan en borrador
  max_respuestas_dia: number;

  activo: boolean;
  creado_por: Nullable<string>;
  fecha_creacion: string;
  fecha_actualizacion: string;
}

export interface APIKey {
  id: string;
  tenant_id: string;
  chatbot_id: string;
  nombre: string;
  key_prefix: string;               // ej: "crb_live_a1b2..." (visible en tabla)
  entorno: EntornoAPIKey;
  activa: boolean;
  
  fecha_expiracion: Nullable<string>;
  ultimo_uso: Nullable<string>;
  
  // Límites
  requests_por_minuto: number;
  requests_por_dia: number;

  creado_por: Nullable<string>;
  fecha_creacion: string;
}

// CORRECCIÓN AQUÍ: Renombrado de APIKeyConPlain a APIKeyGenerada
export interface APIKeyGenerada extends APIKey {
  plain_key: string;                // El token completo visible SOLO al crear
}

// --- DTOs (Payloads para enviar al Backend) ---

export interface CrearChatbotRequest {
  nombre: string;
  tipo: TipoChatbot;
  descripcion?: string;
  
  // Config IA Opcional al crear
  modelo_ia?: string;
  prompt_sistema?: string;
  temperatura?: number;
  max_tokens_respuesta?: number;
}

export interface ActualizarChatbotRequest {
  nombre: string;
  tipo: TipoChatbot;
  descripcion?: string;
  activo: boolean;

  // Config IA
  modelo_ia?: string;
  prompt_sistema?: string;
  temperatura?: number;
  max_tokens_respuesta?: number;

  // Permisos (Scopes)
  puede_leer_reclamos?: boolean;
  puede_responder?: boolean;
  puede_cambiar_estado?: boolean;
  puede_enviar_mensajes?: boolean;
  puede_leer_metricas?: boolean;
  requiere_aprobacion?: boolean;
}

export interface CrearAPIKeyRequest {
  nombre: string;
  entorno: EntornoAPIKey; // 'LIVE' o 'TEST'
}

File: backend/internal/model/chatbot.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// --- CHATBOT ---

type Chatbot struct {
	TenantModel

	Nombre      string     `json:"nombre" db:"nombre"`
	Descripcion NullString `json:"descripcion" db:"descripcion"`
	Tipo        string     `json:"tipo" db:"tipo"`

	// Config IA
	ModeloIA           NullString  `json:"modelo_ia" db:"modelo_ia"`
	PromptSistema      NullString  `json:"prompt_sistema" db:"prompt_sistema"`
	Temperatura        NullFloat64 `json:"temperatura" db:"temperatura"`
	MaxTokensRespuesta NullInt64   `json:"max_tokens_respuesta" db:"max_tokens_respuesta"`

	// Scopes
	PuedeLeerReclamos  bool `json:"puede_leer_reclamos" db:"puede_leer_reclamos"`
	PuedeResponder     bool `json:"puede_responder" db:"puede_responder"`
	PuedeCambiarEstado bool `json:"puede_cambiar_estado" db:"puede_cambiar_estado"`
	PuedeEnviarMensajes bool `json:"puede_enviar_mensajes" db:"puede_enviar_mensajes"`
	PuedeLeerMetricas  bool `json:"puede_leer_metricas" db:"puede_leer_metricas"`

	// Restricciones
	RequiereAprobacion bool            `json:"requiere_aprobacion" db:"requiere_aprobacion"`
	MaxRespuestasDia   int             `json:"max_respuestas_dia" db:"max_respuestas_dia"`
	HorarioActivo      json.RawMessage `json:"horario_activo" db:"horario_activo"`
	SedesPermitidas    json.RawMessage `json:"sedes_permitidas" db:"sedes_permitidas"`

	Activo          bool     `json:"activo" db:"activo"`
	CreadoPor       NullUUID `json:"creado_por" db:"creado_por"`
	UltimaActividad NullTime `json:"ultima_actividad" db:"ultima_actividad"`

	Timestamps
}

// Tipos de chatbot.
const (
	ChatbotAsistenteIA  = "ASISTENTE_IA"
	ChatbotWhatsapp     = "WHATSAPP_BOT"
	ChatbotTelegram     = "TELEGRAM_BOT"
	ChatbotCustom       = "CUSTOM"
)

// --- API KEY ---

type APIKey struct {
	TenantModel
	ChatbotID uuid.UUID `json:"chatbot_id" db:"chatbot_id"`

	Nombre    string `json:"nombre" db:"nombre"`
	KeyPrefix string `json:"key_prefix" db:"key_prefix"`
	KeyHash   string `json:"-" db:"key_hash"` // Nunca se serializa

	Entorno         string          `json:"entorno" db:"entorno"`
	Activa          bool            `json:"activa" db:"activa"`
	FechaExpiracion NullTime        `json:"fecha_expiracion" db:"fecha_expiracion"`
	IPsPermitidas   json.RawMessage `json:"ips_permitidas" db:"ips_permitidas"`
	UltimoUso       NullTime        `json:"ultimo_uso" db:"ultimo_uso"`

	RequestsPorMinuto int `json:"requests_por_minuto" db:"requests_por_minuto"`
	RequestsPorDia    int `json:"requests_por_dia" db:"requests_por_dia"`

	FechaCreacion time.Time `json:"fecha_creacion" db:"fecha_creacion"`
	CreadoPor     NullUUID  `json:"creado_por" db:"creado_por"`
}

// Entornos de API key.
const (
	EntornoLive = "LIVE"
	EntornoTest = "TEST"
)

// --- CHATBOT LOG ---

type ChatbotLog struct {
	TenantModel
	ChatbotID uuid.UUID `json:"chatbot_id" db:"chatbot_id"`
	APIKeyID  uuid.UUID `json:"api_key_id" db:"api_key_id"`

	Metodo      string          `json:"metodo" db:"metodo"`
	Endpoint    string          `json:"endpoint" db:"endpoint"`
	RequestBody json.RawMessage `json:"request_body" db:"request_body"`

	StatusCode   int             `json:"status_code" db:"status_code"`
	ResponseBody json.RawMessage `json:"response_body" db:"response_body"`

	IPAddress  NullString `json:"ip_address" db:"ip_address"`
	DuracionMS NullInt64  `json:"duracion_ms" db:"duracion_ms"`
	ReclamoID  NullUUID   `json:"reclamo_id" db:"reclamo_id"`
	Accion     NullString `json:"accion" db:"accion"`

	FueRateLimited  bool      `json:"fue_rate_limited" db:"fue_rate_limited"`
	FechaExpiracion time.Time `json:"fecha_expiracion" db:"fecha_expiracion"`
	Fecha           time.Time `json:"fecha" db:"fecha"`
}
